### PR TITLE
Caches call to load product variation attributes

### LINKED ISSUE #17120: Potential performance improvement on heavily variable products.
## Prerequisites

- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [x] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [x] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue

1. Create a variable product with variations based on 2 or more attributes - I've used 3 in the example below.
2. Note the ID of the main product
3. In a wp-cli shell, run the following (replacing 42 with your product ID):

```
$ global $wpdb;
$ $p = wc_get_product( 42 );
$ count( $wpdb->queries );
```
Note the number returned here. For me on my test product, it is 32.
```
$ wc_get_product( 42 );
$ count( $wpdb->queries );
```
Note the number returned here. For me on my test product, it is 35.
```
$ wc_get_product( 42 );
$ count( $wpdb->queries );
```
Note the number returned here. For me on my test product, it is 38.

Each time a variable product is retrieved, most information is retrieved through WP APIs and therefore no additional queries need to run to load it again. However, for each attribute set on a variable product a direct DB query is run, the results of which aren't cached in any way. This leads to multiple loads of a product generated additional queries. One query will be run for each attribute associated with variations. In my example above, the product has variations based on 3 attributes, so three queries are run each time the product is loaded. 

Example query (on a product with 256 variations)

```sql
SELECT meta_value FROM wp_postmeta WHERE meta_key = 'attribute_pa_colour' AND post_id IN (1139,1088,1089,1090,1091,1092,1093,1094,1095,1096,1097,1098,1099,1100,1101,1102,1103,1104,1105,1106,1107,1108,1109,1110,1111,1112,1113,1114,1115,1116,1117,1118,1119,1120,1121,1122,1123,1124,1125,1126,1127,1128,1129,1130,1131,1132,1133,1134,1135,1136,1137,1138,1037,1038,1039,1040,1041,1042,1043,1044,1045,1046,1047,1048,1049,1050,1051,1052,1053,1054,1055,1056,1057,1058,1059,1060,1061,1062,1063,1064,1065,1066,1067,1068,1069,1070,1071,1072,1073,1074,1075,1076,1077,1078,1079,1080,1081,1082,1083,1084,1085,1086,1087,986,987,988,989,990,991,992,993,994,995,996,997,998,999,1000,1001,1002,1003,1004,1005,1006,1007,1008,1009,1010,1011,1012,1013,1014,1015,1016,1017,1018,1019,1020,1021,1022,1023,1024,1025,1026,1027,1028,1029,1030,1031,1032,1033,1034,1035,1036,935,936,937,938,939,940,941,942,943,944,945,946,947,948,949,950,951,952,953,954,955,956,957,958,959,960,961,962,963,964,965,966,967,968,969,970,971,972,973,974,975,976,977,978,979,980,981,982,983,984,985,884,885,886,887,888,889,890,891,892,893,894,895,896,897,898,899,900,901,902,903,904,905,906,907,908,909,910,911,912,913,914,915,916,917,918,919,920,921,922,923,924,925,926,927,928,929,930,931,932,933,934)
```

This can hurt performance if a product is accessed multiple times in the same request (especially since multiple queries run each time, and they aren't necessarily particularly "nice" queries either!).

The queries are generated inside `WC_Product_Variable_Data_Store_CPT::read_variation_attributes()`

## Expected/actual behavior

Multiple loads of a product within the same request do not require additional database queries.

### PR DESCRIPTION
**Pull request by rodrigoprimo**
This PR adds product variation attributes to WP cache when they are first loaded to avoid running the same database queries when the same product is loaded multiple times. This cache will be invalidated whenever product attributes are changed.

WP coding standards fixes were also included.
    
Fixes #17120

### CODE REVIEW

**Comment by rvola** [comment]
Hi @rodrigoprimo!
I just saw the problem on a project with **600** products and **2000** variations per product ... It's a hell.

I also worked a fix on my side, which looks a lot like yours.
My question is :
### Why use wp_cache rather than transients?
Indeed, in the Class, the `read_children` method already uses a transitent.
Would not it be wise for the homogeneity of the code to use a similar cache?

In performing my tests, I noticed that the transient saved 3 queries / products.
I say my word: it's really a question about the choice of method, not a questioning of your work.

**Comment by leewillis77** [comment]
Sorry for the delay in replying here - I've been mega busy. I've just checked out this branch and evaluated it against the issues I was originally seeing. I've tested with combinations of the following:

- WooCommerce v3.2.1 (doesn't have this change)
- WooCommerce branch cache-variable-products-queries (this PR)
- Google Product Feed (GPF) extension v7.2.1 (loads the parent product multiple times when outputting variations)
- Google Product Feed (GPF) extension v7.3.0 (Only loads the parent product a single time - ie works around the issue reported on #17120)

The test pulls the product feed from the Google Product Feed extension for a single product containing 112 variations across 3 attributes.

**Please note**: This is a dev box with lots of stuff running that I didn't care to clean out - so don't look /worry about the *actual* numbers, just the direction of travel.

**WooCommerce 3.2.1 + Google Product Feed 7.2.1**
 **746 queries**
This is the effectively the baseline. The product feed extension loads the parent product multiple time when processing a product with variations. On a store with highly-variable products this generates significant numbers of database queries. 

**WooCommerce 3.2.1 + Google Product Feed 7.3.0**
**410 queries**
This version of the GPF extension loads the parent product only once to avoid this issue. This can therefore be looked at as 'target' performance.

**This PR branch + Google Product Feed 7.2.1**
**410 queries**
This is the original "problematic" version of the GPF extension with the patched version of WooCommerce. As we can see from the number, the queries drop to the same level as loading the parent only once even  though we're still loading it multiple times. 

**This PR branch + Google Product Feed 7.3.0**
**410 queries**
Just confirmation that the two versions of everything work together. Performance as expected. 

In summary - I can safely say that this PR resolves the performance issue I originally saw in the Google Product Feed plugin, and will likely have a benefit anywhere that a parent variable product is loaded multiple times in the course of a page request

**Comment by leewillis77** [comment]
Replying to @rvola 

> Why use wp_cache rather than transients?

I can't speak for @rodrigoprimo about why he did it this way.

However, I will note that transients aren't guaranteed to persist, and may disappear at any time without notice. That can lead to unpredictable performance issues which you'd want to avoid. 

To quote the [Codex](https://codex.wordpress.org/Transients_API):

> Everyone seems to misunderstand how transient expiration works, so the long and short of it is: transient expiration times are a maximum time. There is no minimum age. Transients might disappear one second after you set them, or 24 hours, but they will never be around after the expiration time.


My vote (FWIW) would be to stick with cache not transients.

**Comment by rodrigoprimo** [comment]
Just for future reference, I tried to answer why I think cache instead of transients is a better option in this case in the PR (#17327) opened by @rvola.
